v2.1/doc/makedict.1
'\"
'\" Copyright 1998, The University of Texas at Austin ("U. T. Austin").
'\" All rights reserved.
'\"
'\" By using this software the USER indicates that he or she has read,
'\" understood and will comply with the following:
'\"
'\" U. T. Austin hereby grants USER permission to use, copy, modify, and
'\" distribute this software and its documentation for any purpose and
'\" without fee, provided that:
'\"
'\" 1. the above copyright notice appears in all copies of the software
'\"    and its documentation, or portions thereof, and 
'\" 2. a full copy of this notice is included with the software and its
'\"    documentation, or portions thereof, and 
'\" 3. neither the software nor its documentation, nor portions thereof,
'\"    is sold for profit. Any commercial sale or license of this software,
'\"    copies of the software, its associated documentation and/or
'\"    modifications of either is strictly prohibited without the prior
'\"    consent of U. T. Austin. 
'\" 
'\" Title to copyright to this software and its associated documentation
'\" shall at all times remain with U. T. Austin. No right is granted to
'\" use in advertising, publicity or otherwise any trademark, service
'\" mark, or the name of U. T. Austin.
'\" 
'\" This software and any associated documentation are provided "as is,"
'\" and U. T. AUSTIN MAKES NO REPRESENTATIONS OR WARRANTIES, EXPRESSED OR
'\" IMPLIED, INCLUDING THOSE OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR
'\" PURPOSE, OR THAT USE OF THE SOFTWARE, MODIFICATIONS, OR ASSOCIATED
'\" DOCUMENTATION WILL NOT INFRINGE ANY PATENTS, COPYRIGHTS, TRADEMARKS OR
'\" OTHER INTELLECTUAL PROPERTY RIGHTS OF A THIRD PARTY. U. T. Austin, The
'\" University of Texas System, its Regents, officers, and employees shall
'\" not be liable under any circumstances for any direct, indirect, special,
'\" incidental, or consequential damages with respect to any claim by USER
'\" or any third party on account of or arising from the use, or inability
'\" to use, this software or its associated documentation, even if U. T.
'\" Austin has been advised of the possibility of those damages.
'\" 
'\" Submit commercialization requests to: Office of the Executive Vice
'\" President and Provost, U. T. Austin, 201 Main Bldg., Austin, Texas,
'\" 78712, ATTN: Technology Licensing Specialist.
'\"
'\" $Id: makedict.1,v 1.3 1998/07/16 00:00:00 clyde Exp $ (cc.utexas.edu)
'\"
.TH makedict 1
.SH NAME
makedict \- build npasswd dictionaries
.br
packer \- pack hashed dictionary files
.br
unpacker \- unpack hashed dictionary files
.SH SYNOPSIS
@NPASSWD-UTIL@/\fBmakedict\fP
.BI \-\^o " output"
[
.BI \-\^m " source"
] [
.BI \-\^T " temp"
] 
words ... words
.br
@NPASSWD-UTIL@/\fBpacker\fP dictionary-name
.br
@NPASSWD-UTIL@/\fBunpacker\fP dictionary-name
.SH DESCRIPTION
.B Makedict
builds the hashed dictionaries used by
.BR npasswd .
The dictionary hash files are built from \fIword lists\fP, which
are text files with one word per line.
.PP
Three hashed files are created with the suffixes \fB.hwm\fP, \fB.pwd\fP, and \fB.pwi\fP .
.PP
.B Packer
is a low-level utility which reads words (one per line) from standard
input and packs them into hash files.
.B Makedict
is a smart front-end to 
.BR packer ,
and is the preferred tool for dictionary building.
.PP
.B Unpacker
dumps dictionary hashes to standard output, one word per line.
.SH OPTIONS
.PP
Options for
.BR makedict .
.TP
.BI \-\^o " output"
Create the dictionary 
.BR output .
.TP
.BI \-\^m " source"
Merge in words in dictionary
.BR source .
.TP
.BI \-\^T " temp"
Use 
.B temp
as the temporary directory for sorting.  Use this option if the default
temporary directory used by
.IR sort (1)
is small.
.TP
.I words ... words
Word lists. Compressed files (ending with \fB.Z\fP or \fB.gz\fP) can be used.
.PP
.B Packer
and 
.B unpacker
take one argument \- the name of the dictionary file to pack or unpack.
.SH EXAMPLES
.PP
To make the dictionary 'my-dict':
.RS
.nf
makedict -o my-dict word-list-1 ... word-list-n
.fi
.sp
Will produce the files \fBmy-dict.pwd\fP,
\fBmy-dict.pwi\fP and \fBmy-dict.hwm\fP.
.RE
.PP
To add words to an existing dictionary:
.RS
.nf
makedict -o existing-dict -m existing-dict word-list-1 ... word-list-n
.fi
.RE
.SH DIAGNOSTICS
.TP 5
Missing option to -o
No output file was specified.
.TP 5
Missing option to -T
No temporary directory was specified.
.TP 5
Missing option to -m
No merge dictionary was specified.
.TP 5
Empty or missing hash file \fIname\fP
One of the dictionary hash files came up empty, either because the
sources were empty or a filesytem error occurred.
.SH BUGS
The input to
.B packer
.B must
.B be
.BR sorted ,
else the result hash files will be faulty.  Use 
.B makedict
instead.
.SH "SEE ALSO"
npasswd(1)
.br
\fINpasswd Reference Manual\fP
.SH AUTHOR
Clyde Hoover
.br
Academic Computing Services and Instructional Technology Services
.br
The University of Texas at Austin
.br
c.hoover@cc.utexas.edu
.br
\(co 1998, The University of Texas at Austin. All rights reserved. 
'\"
'\" End $RCSfile: makedict.1,v $


v2.1/NEWS
Latest npasswd news	$Date: 1999/02/01 00:00:00 $

Patch 05	2/1/99	A passel of little bug fixes
	* Added copyright messages to more modules.
	* A bit of code format cleanup.
	* Do not enable nsswitch support unless the "nsswitch.h" include
	  file is present.  This means that nsswitch.conf will ONLY be
	  supported on Solaris 2.
	  IRIX 6.5 and Redhat Linux 4.2 have /etc/nsswitch.conf but totally
	  different API for support.

	* Properly punt the '-r' option on SunOS 5.
	* Fixed minor Makefile things throughout the distribution.
	* Added the option "-XA" to always prompt for the old password.
	* Fixed lexical password check to properly recoginize telephone numbers.
	* Enhanced debug mode in checkpassword program.
	* Use an internal pager to present help files.
	* Reorganize terminal handling code.
	* Fix -XI option to work right when stdin is a pipe
	* Fix Configure password file list input loop
	* Re-activate some Cracklib rules to be more anal about password
	  checking.

Patch 04	10/15/98	Lots of bug fixes, some feature enhacements
	* Substantial rework of history_admin:
		* Add 'merge' function (see manual page)
		* Add '-i' command line option 
		* Allow '-m' with default database
		* Fix some more annoying bugs
	* Modifications to the password history test suite to use
	  and verify the above changes
	* Clarify some issues with password history depth
		* History depth now means "how many *old* passwords matter"
		* Always retain the more recent password, even if too old
	* Allow "@" to mean the default database name in history
	  configuration directives (and the history_admin '-f' option)
	* Fix the "passwd.MaxRepeat" config directive
	* Fix the "ShadowTolerance" config directive
	* Fix the "PasswdTolerance" config directive
	* Fix the "PasswdMatchTries" config directive
	* Add build-time option "SHORT_FILE_WARN" to control behavior
	  when passwd/shadow temp files come up short
	* Support the '-r' option under SunOS 5
	* Allow change of local passwd even if NIS client (unless SunOS 4)
	* More run-time trace messages in various places

Patch 03	9/24/98
	* Fixes the "Cannot talk to NIS server" problem under Solaris 2. 

Patch 02	9/22/98
	* Fixes the build problems on Digital UNIX 4.0d.

Patch 01	9/18/98
	* Fixes build problems when NIS support is not enabled.
	* Fixes a problem with the configuration directive "PrintOnly", which
	  was renamed "PrintableOnly" during the last development cycle but
	  the change did not get propigated everywhere.

Platform status:

AIX 4.1		Builds, installs, changes local passwords
SunOS 4.1.4	Builds, installs, changes local and NIS passwords 
SunOS 5.5	Builds, installs, changes local and NIS passwords 
IRIX 6.1	Builds, installs, changes local passwords 
Digital UNIX 4.0	Builds, installs



V2.0/src/PasswordCheck/cracklib/tools/makedict.SH
case $CONFIG in
'')
	if test -f config.sh; then TOP=.;
	elif test -f ../config.sh; then TOP=..;
	elif test -f ../../config.sh; then TOP=../..;
	elif test -f ../../../config.sh; then TOP=../../..;
	elif test -f ../../../../config.sh; then TOP=../../../..;
	else
		echo "Can't find config.sh."; exit 1
	fi
	. $TOP/config.sh
	;;
esac
: This forces SH files to create target in same directory as SH file.
: This is so that make depend always knows where to find SH derivatives.
case "$0" in
*/*) cd `expr X$0 : 'X\(.*\)/'` ;;
esac
echo "Extracting makedict (with variable substitutions)"
: This section of the file will have variable substitutions done on it.
: Move anything that needs config subs from !NO!SUBS! section to !GROK!THIS!.
: Protect any dollar signs and backticks that you do not want interpreted
: by putting a backslash in front.  You may delete these comments.
$spitshell >makedict <<!GROK!THIS!
$startsh

TR=$tr
TRUP="$trup"
TRDOWN="$trdown"
PACKER=$p_utildir/packer	# The program that packs the dictionary
UNPACKER=$p_utildir/unpacker	# The program that unpacks the dictionary

!GROK!THIS!

: In the following dollars and backticks do not need the extra backslash.
$spitshell >>makedict <<'!NO!SUBS!'
#
# @(#)makedict.SH	1.6 07/16/98 (cc.utexas.edu)
#
###
# This program is copyright Alec Muffett 1993. The author disclaims all
# responsibility or liability with respect to it's usage or its effect
# upon hardware or computer systems, and maintains copyright as set out
# in the "LICENCE" document which accompanies distributions of Crack v4.0
# and upwards.
###
#
# This script is used to build the dictionary hash files
#
# Usage:
#	mkdict -o output-dict
#		[-T /tmp/dir (alternate tmp dir)]
#		[-m merge-with-dict]
#		dict-source1 dict-source2 ...
#
### In case of explosion, rerun with with "-T" option pointing to a lot
### of free space in a directory somewhere.

Usage="Usage: $0 -o output-dict [-m source-dict] [-T /tmp/dir] dict-source1 dict-source2 ..."
SORT="sort"

if [ $# -lt 1 ]; then
	echo $Usage
	exit 1
fi
while [ -n "$1" ]; do
	case $1 in
	-m*)	src_dict=`expr $1 : '-s\(.*\)'`
		if [ -z "$src_dict" ]; then
			shift
			src_dict=${1?"Missing option to -m"}
		fi
		;;

	-o*)	dest_dict=`expr $1 : '-o\(.*\)'`
		if [ -z "$dest_dict" ]; then
			shift
			dest_dict=${1?"Missing option to -o"}
		fi
		;;

	-T*)	tmp=`expr $1 : '-T\(.*\)'`
		if [ -z "$tmp" ]; then
			shift
			tmp=${1?"Missing option to -T"}
		fi
		;;

	*)	break
		;;
	esac
	shift
done
#
# Add the directory where this program lives to PATH
# so that 'packer' will be found (if needed)
#
if [ ! -f $PACKER ]; then
	bn=`basename $0`
	pn=`expr $0 : "\(.*\)/$bn"`
	PACKER="$pn/$PACKER"
	UNPACKER="$pn/$UNPACKER"
fi

if [ -z "$dest_dict" ]; then
	echo No output file given
	echo $Usage
	exit 1
fi
[ -n "$tmp" ] && SORT="$SORT -T $tmp"

OUT=${dest_dict}.TMP
(
	[ -n "$src_dict" ] && $UNPACKER $src_dict
	for f in $*; do
		for x in '' .Z .gz ''; do
			if [ -f "$f$x" ]; then
				f=$f$x
				break
			fi
		done
		case "$f" in
		*.Z)	zcat $f;;
		*.gz)	gunzip --stdout $f;;
		*)	cat $f;;
		esac
	done
) |	grep -v '^#' | \
	grep -v '^$' | \
	$TR "$TRUP" "$TRDOWN" | \
	$SORT | \
	uniq | \
	$PACKER $OUT		# Build the dictionary

for t in hwm pwd pwi; do
	if [ -s $OUT.$t ]; then
		mv $OUT.$t $dest_dict.$t
	else
		echo Empty or missing hash file $OUT.$t
	fi
done
echo Made dictionary $dest_dict
exit 0
#
# End makedict.SH
!NO!SUBS!
chmod 755 makedict
$eunicefix makedict
